### ISSUE TITLE
Non-recursive callbacks for BIOS and DOS functions

### ISSUE THREAD

**Issue by joncampbell123** [joncampbell123/dosbox-x#919]
**Is your feature request related to a problem? Please describe.**
- Some parts, including the command shell and BIOS emulation, access memory or use the normal loop in a way that involves recursion of the emulation loop (and CBRET_STOP)

**Describe the solution you'd like**
- Reduce BIOS and DOS emulation functions where possible to callback routines that do minimal native code, where the native code can operate with the guarantee not to recurse into the emulation loop.
- Accomplish this through writing ASM in an external assembler, and placing the final binary code into memory in DOSBox-X and pointing vectors at them.

**Additional context**
- In real mode, I/O and memory access are trivially non-recursive.
- In vm86 mode, if paging is enabled, memory access can cause a page fault. Normal emulation can handle a page fault, a callback function in native code cannot (at least the way they're written now) without recursing into an emulation loop to handle the page fault.
- In protected mode, I/O port access can cause an exception. Normal emulation can handle it, callback functions cannot without recursing into an emulation loop.
- DOS emulation (such as the console) also uses recursion to call into the BIOS INT 10h interrupt when printing to screen, or the command shell which is a callback function to then call BIOS and DOS functions in a loop to respond.
- Non-recursive emulation is REQUIRED to get DOSBox-X to run in the browser via emscripten without terrible hacks or use of runtime interpretation (would help @yksoft1 and his branch targeting the browser). Emscripten compiled code runs fastest when the C code runs as JS or WASM, but with the condition the emulation must yield to the web browser after each millisecond.
- The idea is to write portions of BIOS and DOS emulation in ASM, assemble with NASM (to produce OBJ files), then convert the OBJ files to unsigned char[] (the binary) and const size_t variables (one for each global label) in a header that DOSBox-X code can copy into guest memory and patch up using the labels to make it work.
- As guest-visible executable x86 code, DOS and BIOS interrupt handling can be safely run in vm86 by Windows, I/O trapped by protected mode, and preempted by the task scheduler in Windows 95, without requiring the normal loop support recursing into itself. The command shell in this form can do the work of talking to the DOS shell and interacting through x86 code in the guest, and leave the shell command processing to native code when the user hits enter. The DOS CON device emulation can become x86 code in the guest where it can then call into the BIOS without recursing into the normal loop.

**Comment by joncampbell123** [comment]
Feedback and suggestions are welcome.

**Comment by joncampbell123** [comment]
Also note that by using OBJ as an intermediary, the guest code could be ASM assembled with NASM, or could be plain C compiled with Open Watcom C++.

I would like to write a lot of the 16-bit code in C with some ASM, but Open Watcom C++ isn't as easy to incorporate into the build as a simple precompiled binary of NASM.

However since the C code runs in a low level environment, the whole C runtime usually associated with Open Watcom doesn't need to be there, and we're not using the Open Watcom linker, so it might be easy to incorporate JUST the compiler executable into the source tree alongside NASM with a very minimal set of headers for it to use.

**Comment by yksoft1** [comment]
Do you mean to bring a special port of FreeDOS with binary port of frontends for some DOSBox-specific tools (MOUNT, IMGMOUNT, etc.) into DOSBox-X to replace the current emulated DOS shell ?

**Comment by joncampbell123** [comment]
Portions taken from FreeDOS will more likely replace some commands like COPY.EXE.

I might add an option to use FreeDOS's command shell as well, however by default it would be the DOSBox/DOSBox-X command shell implemented in ASM running in DOSBox-X.

**Comment by joncampbell123** [comment]
The idea is that the command shell and other utilities would be written in ASM and assembled into binary that DOSBox-X runs as COMMAND.COM, but the asm would have some callback instructions into DOSBox-X for anything that can be emulated in native code as long as it doesn't involve recursion into the normal loop.

**Comment by joncampbell123** [comment]
Callbacks are allocated dynamically, which is why the proposed change says to set fixed values at 0x8000 or higher (bit 15 set) for the callbacks these binaries will need.

**Comment by yksoft1** [comment]
It seemed to be a really big work that should be considered as a long-term objective.. 
I don't think including a compiler or assembler binary into the repository is any good, given that won't work cross-platform. It's better to just require nasm/yasm in build scripts.

**Comment by joncampbell123** [comment]
I'm sorry I didn't make it clear, that's what I intend to do. If the ASM files backing the command shell and other components are modified, then NASM is used to generate executable x86 code that is stored in DOSBox-X and then copied into guest memory when needed, and then executed by the emulator as x86 opcodes running in a DOS machine.

I'm well aware using NASM for native code on the host is very non-portable. I can think of several emulators that are unusable already on x86_64 because of that, such as a certain SNES emulator.

What the 16-bit asm will be allowed to do is use the DOSBox callback instruction with callback numbers specifically set aside for the matching function in DOSBox-X (with bit 15 set, as listed).

**Comment by yksoft1** [comment]
I acknowledge what you will do is to replace the emulated DOS shell and commands using binary blobs, not to use assembly in emulator core (cpu/hardware/ints/bios) itself. But compiling such binary blobs would involve cross-compiling/cross-assembling on non-x86 hosts that could make the build envionment more complicated.

**Comment by joncampbell123** [comment]
Not really. NASM is a simple assembler that is cross-platform and open source. It can be found on most Linux distributions already, and precompiled EXE files for Windows can be found as well.

I also want to be able to compile from non-x86 hosts, so that I can develop DOSBox-X from a Raspberry Pi 3 I have sitting around, for example (ARMv7 based).

**Comment by yksoft1** [comment]
> Not really. NASM is a simple assembler that is cross-platform and open source. It can be found on most Linux distributions already, and precompiled EXE files for Windows can be found as well.
> 
> NASM is very simple to compile from source, and it doesn't need much at runtime to work.
> 
> I also want to be able to compile from non-x86 hosts, so that I can develop DOSBox-X from a Raspberry Pi 3 I have sitting around, for example (ARMv7 based).

So it's just modifying configure.am and makefile.am shit to check nasm/yasm on the host. AFAIK either nasm or yasm only process x86/x86-64 assembly and generate x86/x86-64 code, so there is no need to check whether nasm/yasm is a cross-assembler.

**Comment by joncampbell123** [comment]
I'm only interested in using NASM to generate x86 code, regardless of the host CPU, so that's perfectly fine.

I'm going to use NASM to generate x86 code that will run in the emulated DOSBox environment. It doesn't matter what the host is.

**Comment by joncampbell123** [comment]
This is a long-term idea that will not happen immediately, it will happen over time. Emulation will only get worse if I do this in haste, so I will do it carefully and slowly.

It will help with targeting Emscripten and the web browser, and it will help with protected mode/virtual 8086 mode virtualization of BIOS and DOS functions.

**Comment by joncampbell123** [comment]
I think the best way to add it in is to write it so that if you do not have NASM, DOSBox-X can compile with the already generated output anyway. You'll only need NASM if you want to modify those parts. I think that might be the best compromise.

**Comment by joncampbell123** [comment]
I just completed a few changes to the 16-bit linker I wrote in DOSLIB to accomplish this ASM task.

The changes allow emitting the compiled binary to an unsigned char[] array in a CPP file and extern defs to it in a header.

The header also contains #defines for each segment and symbol that can be used by the native DOSBox-X code to patch up the binary as needed once loaded into guest x86 memory.

16-bit linker is here:
https://github.com/joncampbell123/doslib/tree/master/tool/linker

16-bit linker main executable is defined by this C file:
https://github.com/joncampbell123/doslib/blob/master/tool/linker/lnkdos16.c

It uses an OMF parsing library I wrote here:
https://github.com/joncampbell123/doslib/tree/master/fmt/omf

Example output here (one COM, one EXE):
https://github.com/joncampbell123/doslib/tree/master/tool/linker/example-hexsplit-obj

Proposed build steps:
- ASM -> NASM -> OBJ (OMF format)
- One or more OBJ -> LNKDOS16 -> BINARY + H + CPP file
- H + CPP file are compiled into DOSBox-X
- DOSBox-X code copies binary into guest x86 memory, then patches as necessary to connect the code together with other binary blobs.

All binary blobs will be open source as DOSBox-X is. You will only need NASM and LNKDOS16. Both are open source, light on compile time and runtime dependencies, and easy to put into build steps. Both can be used in builds steps on systems other than x86. Both can be placed as compiled EXEs in the tree for use on Windows.

**Comment by joncampbell123** [comment]
configure.ac now looks for NASM and the LNKDOS16 linker in DOSLIB.

Support will be added in a way that, if either is not available, should not prevent DOSBox-X from compiling unless you modify the .ASM files.
